File: Decorator/c++/Decorator.cpp
#include <iostream>
#include <string>

class Processor
{
    public:
        virtual void process() = 0;
};

class Transmitter : public Processor
{
    private:
        std::string data;

    public:
        Transmitter(std::string data) : data(data) {}
        void process() override;
};

class Shell : public Processor
{
    protected:
        Processor* processor;

    public:
        Shell(Processor* processor) : processor(processor) {}
        void process() override;
};

class HammingCoder : public Shell
{
    public:
        HammingCoder(Processor* processor) : Shell(processor) {}
        void process() override;
};

class Encryptor : public Shell
{
    public:
        Encryptor(Processor* processor) : Shell(processor) {}
        void process() override;
};

int main()
{
    Processor* transmitter = new Transmitter("12345");
    transmitter->process();
    std::cout << std::endl;

    Shell* hammingCoder = new HammingCoder(transmitter);
    hammingCoder->process();
    std::cout << std::endl;

    Shell* encryptor = new Encryptor(hammingCoder);
    encryptor->process();
    
    return 0;
}

void Transmitter::process()
{
    std::cout << "Data " << data << " transmitted via the communication channel\n";
}

void Shell::process()
{
    processor->process();
}

void HammingCoder::process()
{
    std::cout << "noise-resistant Hamming code has been applied->";
    processor->process();
}

void Encryptor::process()
{
    std::cout << "data encryption->";
    processor->process();
}

File: Decorator/python/Decorator.py
import abc


class IProcessor(metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def process(self):
        pass


class Transmitter(IProcessor):

    def __init__(self, data: str):
        self.__data = data

    def process(self):
        print('Data {} transmitted via the communication channel'.format(self.__data))


class Shell(IProcessor):

    def __init__(self, pr: IProcessor):
        self._processor = pr

    @abc.abstractmethod
    def process(self):
        self._processor.process()


class HammingCoder(Shell):

    def __init__(self, pr: IProcessor):
        super().__init__(pr)

    def process(self):
        print('noise-resistant Hamming code has been applied->', end='')
        self._processor.process()


class Encryptor(Shell):

    def __init__(self, pr: IProcessor):
        super().__init__(pr)

    def process(self):
        print('data encryption->', end='')
        self._processor.process()


if __name__ == '__main__':

    transmitter: IProcessor = Transmitter('12345')
    transmitter.process()
    print()

    hamming_coder: Shell = HammingCoder(transmitter)
    hamming_coder.process()
    print()

    encryptor: Shell = Encryptor(hamming_coder)
    encryptor.process()